R/pkg_ref_class.R

```diff
@@ -122,69 +141,110 @@ as_pkg_ref.pkg_ref <- function(x, ...) {
 
 #' @importFrom utils installed.packages available.packages packageVersion
 #' @export
-as_pkg_ref.character <- function(x, repos = getOption("repos"), ...) {
-  ip <- memoise_installed_packages()
+as_pkg_ref.character <- function(x, repos = getOption("repos", "https://cran.rstudio.com"),
+                                 source = NULL, lib.loc = NULL, ...) {
 
-  # case when only a package name is provided
-  #   e.g. 'dplyr'
-  # regex from available:::valid_package_name_regexp
-  if (grepl("^[[:alpha:]][[:alnum:].]*[[:alnum:]]$", x)) {
+  dots <- list(...)
 
-    # if locally installed
-    if (x %in% ip[,"Package"]) {
-      path <- find.package(x)
-      version <- utils::packageVersion(x, lib.loc = dirname(path))
+  pkg_source_ <- ifelse(is.null(source),
+                        determine_pkg_source(x, source, repos),
+                        source)
+
+  stopifnot(pkg_source_ %in% c("pkg_install", "pkg_source", "pkg_cran_remote",
+                               "pkg_bioc_remote", "pkg_missing"))
+
+  switch(pkg_source_,
+         pkg_install = {
+           path <- find.package(x, lib.loc = lib.loc)
+           version <- utils::packageVersion(x, lib.loc = dirname(path))
+
+           new_pkg_ref(x,
+                       version = version,
+                       path = path,
+                       source = "pkg_install")
+         },
+         pkg_source = {
+
```

Comment:
Could we provide a more informative error message when `source` is not available? 

Below is what currently have. 

```
> ref5 <- pkg_ref("testthat", source = "pkg_source")
Error in read.dcf(file.path(x, "DESCRIPTION")) : 
  cannot open the connection
In addition: Warning message:
In read.dcf(file.path(x, "DESCRIPTION")) :
 
 Error in read.dcf(file.path(x, "DESCRIPTION")) : 
  cannot open the connection 
```

Reply:
Similarly, I feel we shall we report warning/error if a package is not in a source? 

```
> ref5 <- pkg_ref("testthat2", source = "pkg_cran_remote")
> ref5
<pkg_cran_remote, pkg_ref> testthat2 v
$repo
  character(0)
$source
  [1] "pkg_cran_remote"
$version
  character(0)
$name
  [1] "testthat2"
$archive_release_dates...
$bug_reports...
$bug_reports_host...
$bug_reports_url...
$downloads...
$license...
$news_urls...
$source_control_url...
$web_url...
$website_urls...
```

Reply:
I've added nice warnings for these and change the source type to 'pkg_missing". Tests have been added to check for this behavior